src/lib/operators/projection.cpp

```diff
@@ -290,4 +295,45 @@ std::shared_ptr<Table> Projection::dummy_table() {
   return shared_dummy;
 }
 
+/**
+ *  Method to determine PQPColumns to forward. In most cases, forwarding a column comes with almost zero costs.
+ *  However, in cases where the same PQPColumn is also accessed in the expression evaluation, the whole column is
+ *  materialized and cached anyways. In that case, returning the materialized PQPColumn instead of forwarding it
+ *  benefits the following operator which can process a fully materialized column sequentially instead of accessing the
+ *  column indirectly via a position list.
+ */
+ExpressionUnorderedSet Projection::_determine_forward_columns(const TableType table_type) const {
+  // We collect all forwardable PQP column expressions and all evaluated PQP column expressions. In a second step, all
+  // forwardable expressions that are also evaluated, are removed from the set.
+  auto forwardable_pqp_columns = ExpressionUnorderedSet{};
+  auto evaluated_pqp_columns = ExpressionUnorderedSet{};
+  for (auto column_id = ColumnID{0}; column_id < expressions.size(); ++column_id) {
+    const auto& expression = expressions[column_id];
+
+    if (expression->type == ExpressionType::PQPColumn) {
+      forwardable_pqp_columns.emplace(expression);
+      continue;
+    }
+
+    visit_expression(expression, [&](const auto& sub_expression) {
+      if (const auto pqp_column_expression = std::dynamic_pointer_cast<PQPColumnExpression>(sub_expression)) {
+        evaluated_pqp_columns.emplace(sub_expression);
+      }
+      return ExpressionVisitation::VisitArguments;
+    });
+  }
+
+  // Do not forward columns which are also evaluated. Not relevant for data tables as the following operator does not
+  // need to iterate via a position list anyways. In cases where the data table is encoded (e.g., dictionary encoded)
+  // and the following operator has fast paths for dictionary-encoded segments (e.g., hash aggreagate), forwarding over
+  // using the materialized column of the expression evaluator can be beneficial.
```

Comment:
This somewhat contradicts the previous explanation. The limitation on reference segments should be described there, too.

Reply:
Changed. Can you take a look at it?
Code changed so did not paste the comment here.